update github.com/imdario/mergo to v0.3.5

cblecker opened the pull request:
**What this PR does / why we need it**:
Updates github.com/imdario/mergo library to v0.3.5. We were pinned because of a functionality change in the dependency, however, a new function was introduced with similar functionality to the old.

There is apparently some Debian packaging issues (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=878254) because of this. I'm still not clear what those are, unless they are forcing the library to update, as opposed to using our `vendor/`.

That said, this will allow for some of those vendor conflicts to resolve for anyone else who is using client-go, so that's at least worthwhile.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
fixes #27543, fixes https://github.com/kubernetes/client-go/issues/431

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

cblecker committed:
Update github.com/imdario/mergo to v0.3.5

cblecker commented:
/test pull-kubernetes-kubemark-e2e-gce-big

cblecker commented:
/assign @deads2k @sttts 
The tests are passing here, but I'm not fluent on this part of the code base. If the tests give appropriate coverage, then we might be good to merge. PTAL :)

sttts commented:
> That said, this will allow for some of those vendor conflicts to resolve for anyone else who is using client-go, so that's at least worthwhile.

Debian uses *-dev packages which enforce a base line of all Golang dependencies throughout a system. My first thought: insane, and not supportable by us in any way.

cblecker commented:
@sttts Oi. Yeah, I agree. That said, making this change could also possibly solve vendor conflicts for those tho might import client-go, but also use mergo somewhere else in their codebase.

If it’s safe to make, I think it’s worth making. If it’s not safe, then I’ll happily close.

sttts commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config_test.go:
```diff
@@ -34,7 +34,7 @@ func TestOldMergoLib(t *testing.T) {
 	dst := T{X: "one"}
 	src := T{X: "two"}
-	mergo.Merge(&dst, &src)
+	mergo.MergeWithOverwrite(&dst, &src)
```
this test was added to warn us when somebody revendors mergo in a new version. At least we should rename the test (I would keep it though), maybe: TestMergoSemantics. It's not about the old mergo anymore after this PR.

sttts replied:
Also update the comment below.

sttts replied:
Actually, when seeing the many changes, having some more complex smoke testing here wouldn't harm, i.e. all non-simple types like maps and slices. Then please test those before the bump of mengo and after.

cblecker committed:
Move from mergo.Merge to mergo.MergeWithOverwrite

cblecker commented:
Updated test:
```
$ bazel test //staging/src/k8s.io/client-go/tools/clientcmd:go_default_test --test_output=all --test_arg=-test.v --test_filter=TestMergoSemantics
INFO: Build options have changed, discarding analysis cache.
INFO: Analysed target //staging/src/k8s.io/client-go/tools/clientcmd:go_default_test (0 packages loaded).
INFO: Found 1 test target...
INFO: From Testing //staging/src/k8s.io/client-go/tools/clientcmd:go_default_test:
==================== Test output for //staging/src/k8s.io/client-go/tools/clientcmd:go_default_test:
=== RUN   TestMergoSemantics
--- PASS: TestMergoSemantics (0.00s)
PASS
================================================================================
Target //staging/src/k8s.io/client-go/tools/clientcmd:go_default_test up-to-date:
  bazel-bin/staging/src/k8s.io/client-go/tools/clientcmd/darwin_amd64_stripped/go_default_test
INFO: Elapsed time: 1.409s, Critical Path: 0.65s
INFO: 1 process: 1 darwin-sandbox.
INFO: Build completed successfully, 2 total actions
//staging/src/k8s.io/client-go/tools/clientcmd:go_default_test           PASSED in 0.3s

Executed 1 out of 1 test: 1 test passes.
INFO: Build completed successfully, 2 total actions
```

cblecker committed:
Improve TestMergoSemantics test

cblecker commented:
@sttts PTAL. I added a test for nested structs, and refactored it to allow us to add more scenarios. I'm not sure what other kinds of scenarios would be useful to exercise, based on how we use this. If you have any specific examples you're thinking about, I can easily add it as a new case.

sttts commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config_test.go:
```diff
@@ -28,21 +28,48 @@ import (
+	var testData = []struct {
+		dst      T
+		src      T
+		expected T
```
how about slices and maps? Those are the most interesting for merge.

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -211,22 +211,22 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
+		mergo.MergeWithOverwrite(nonMapConfig, kubeconfig)
 	}
 
 	// since values are overwritten, but maps values are not, we can merge the non-map config on top of the map config and
```
> since values are overwritten, but maps values are not

is this still true with the new version?

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -211,22 +211,22 @@ func (rules *ClientConfigLoadingRules) Load() (*clientcmdapi.Config, error) {
+		mergo.MergeWithOverwrite(mapConfig, kubeconfig)
 	}
 
 	// merge all of the struct values in the reverse order so that priority is given correctly
```
> merge all of the struct values in the reverse order so that priority is given correctly

is this still true with the new version?

liggitt commented:
according to the report, 0.3.5 would fix https://github.com/kubernetes/client-go/issues/431 as well

liggitt commented:
> There is apparently some Debian packaging issues (https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=878254) because of this.

that's particularly concerning, if it means they are building with a custom set of vendored components... that means they could be producing a binary with completely different behavior than what is tested/released

liggitt commented:
/retest
/lgtm

liggitt commented:
/hold

for @sttts comment about tests for slices/maps
of particular interest would be a test exercising https://github.com/kubernetes/client-go/issues/431

liggitt commented:
/retest

cblecker commented:
@liggitt @sttts Added some more basic tests for merging maps and merging slices as a part of a struct. It also looks like we have other tests that cover the functions that we use mergo in. I think this should be safe to merge at this point.

Also added the test case from https://github.com/kubernetes/client-go/issues/431.

cblecker committed:
add more tests to clientcmd

cblecker commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config_test.go:
```diff
@@ -553,3 +588,46 @@ func TestNamespaceOverride(t *testing.T) {
 	matchStringArg("foo", ns, t)
 }
+
+func TestAuthConfigMerge(t *testing.T) {
```
this passes on master without the mergo bump... can we add a test that reproduces https://github.com/kubernetes/client-go/issues/431 on master?

cblecker replied:
This is the exact test from that issue, but I can't actually reproduce https://github.com/kubernetes/client-go/issues/431 at this time. Should I just remove this test completely?

liggitt replied:
turns out this test caught issues only present between 0.3.0 and 0.3.2. it's fine to keep it anyway

ncdc commented:
/uncc

liggitt commented:
/lgtm
/hold cancel

cblecker commented:
/test pull-kubernetes-e2e-kops-aws

mrshelly commented:
??
Thank you , For read this Email !

________________________________
发件人: Christoph Blecker <notifications@github.com>
发送时间: 2018年8月12日 0:48
收件人: kubernetes/kubernetes
抄送: Subscribed
主题: Re: [kubernetes/kubernetes] update github.com/imdario/mergo to v0.3.5 (#65779)


@liggitt<https://github.com/liggitt> @sttts<https://github.com/sttts> Added some more basic tests for merging maps and merging slices as a part of a struct. It also looks like we have other tests that cover the functions that we use mergo in. I think this should be safe to merge at this point.

Also added the test case from kubernetes/client-go#431<https://github.com/kubernetes/client-go/issues/431>.

―
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/kubernetes/kubernetes/pull/65779#issuecomment-412310835>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAEfRKsfPDjtlMt1nZAlnEfPoCibR5D3ks5uP3tXgaJpZM4VBbw2>.
